README.md
This is a small library that validates HTTP requests and responses against an OpenAPI 3.0 specification.

It is designed to work with any web servers/framework that uses the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) such as:

* [Cloudflare Workers](https://developers.cloudflare.com/workers/)
* [Astro](https://astro.build/)
* [Remix](https://remix.run/)

It also provides adapters for non-Fetch based web servers/frameworks such as:

* [AWS Lambda](https://aws.amazon.com/lambda/)
* [Express](https://expressjs.com/) (coming soon)
* [Fastify](https://www.fastify.io/) (coming soon)

The library is built on top of [Zod](https://zod.dev/) and [zod-to-openapi](https://github.com/asteasolutions/zod-to-openapi).

- [Installation](#installation)
- [Define an OpenAPI route](#define-an-openapi-route)
- [Create a validator](#create-a-validator)
- [Validate requests and responses](#validate-requests-and-responses)
- [Error handling](#error-handling)
- [Adapters](#adapters)
  - [AWS Lambda](#aws-lambda)
  - [Express](#express)
  - [Fastify](#fastify)

## Installation

```bash
npm install --save @oselvar/openapi-validator
```

## Define an OpenAPI route

```typescript
import { RouteConfig } from '@asteasolutions/zod-to-openapi';
import { z } from 'zod';

import {
  Response404,
  Response415,
  Response422,
  Response500,
} from '@oselvar/openapi-validator';

// Define Zod schemas for the request parameters, query and body

const ThingParamsSchema = z.object({
  thingId: z.string().regex(/[\d]+/),
});

const ThingQuerySchema = z.object({
  thingId: z.string().regex(/[\d]+/).optional(),
});

const ThingBodySchema = z.object({
  name: z.string().regex(/[a-z]+/),
  description: z.string().regex(/[a-z]+/),
});

// Define an OpenAPI route using https://github.com/asteasolutions/zod-to-openapi

const routeConfig: RouteConfig = {
  method: 'post',
  path: '/things/{thingId}',
  request: {
    params: ThingParamsSchema,
    query: ThingQuerySchema,
    body: {
      content: {
        'application/json': {
          schema: ThingBodySchema,
        },
      },
    },
  },
  responses: {
    200: {
      description: 'Create a thing',
      content: {
        'application/json': {
          schema: ThingBodySchema,
        },
      },
    },
    404: Response404,
    415: Response415,
    422: Response422,
    500: Response500,
  },
};
```

## Create a validator

The `Validator` class uses the schemas in the `routeConfig` object to validate requests and responses.

```typescript
import { Validator } from '@oselvar/openapi-validator';

const validator = new Validator(routeConfig);
```

## Validate requests and responses

Your web server will provide a `request` object and a `params` object.

(If your web server does not use the Fetch API, see the [Adapters](#adapters) section below.)

```typescript
const request: Request = ...;
const params: Record<string, string | undefined> = ...;

const params = validator.params<z.infer<typeof ThingParamsSchema>>(params);
const body = await validator.body<z.infer<typeof ThingBodySchema>>(request);
const response = validator.validate(Response.json(body));
```

## Error handling

The methods on the `Validator` object will throw a `ValidationError` if the request or response is invalid.
You must handle this error and return an appropriate HTTP response to the client.

We recommend doing this in a middleware function. Please refer to your web server's documentation for more information.

Here is an example:

```typescript
import { unwrapError } from '@oselvar/openapi-validator';

try {
  // Run the handler
} catch (error) {
  const { message, response } = unwrapError(error);
  console.error(response.status, message);
  return response;
}
```

## Adapters

If you are using a framework that does *not* use the Fetch API [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) and [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) objects
such as [AWS Lambda](https://aws.amazon.com/lambda/), [Express](https://expressjs.com/) or [Fastify](https://www.fastify.io/), use the `FetchRoute` type to define your handler function:

```typescript
import { FetchRoute, Validator } from '@oselvar/openapi-validator';
import { RouteConfig } from '@asteasolutions/zod-to-openapi';
import { z } from 'zod';

const routeConfig: RouteConfig = { ... };
const validator = new Validator(routeConfig);

const fetchRoute: FetchRoute = async (context) => {
  const params = validator.params<z.infer<typeof ThingParamsSchema>>(context.params);
  const body = await validator.body<z.infer<typeof ThingBodySchema>>(context.request);
  return validator.validate(Response.json(body));
};
```

Note that the support for multiple HTTP servers can also simplify your developer experience.
You can write your handler function once and then register it with multiple HTTP servers.

For example, you can register your handler functions with Express or Fastify during development and then register them with AWS Lambda during production.

The next step is to convert the `FetchRoute` to a function that can be registered with your HTTP server.

### AWS Lambda

```typescript
import { FetchOpenAPIHandler } from '@oselvar/openapi-validator';
import { toProxyHandler } from '@oselvar/openapi-validator/aws-lambda';

export const handler = toProxyHandler(fetchRoute);
```

### Express

Coming soon.

### Fastify

Coming soon.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.1] - 2023-11-27
### Fixed
- Allow responses without a content type (for example redirects).

## [0.3.0] - 2023-11-17
### Added
- Support for `application/x-www-form-urlencoded`
- Export `HttpError` class

## [0.2.1] - 2023-11-16
### Added
- Better support for error responses with `unwrapError`.

## [0.2.0] - 2023-11-16
### Changed
- Added a `validate` method for validating responses.

## [0.1.0] - 2023-11-16
### Changed
- Renamed to `@oselvar/openapi-validator`.
- Refactored the code to use a `Validator` class to validate objects independently.
- Thrown errors have to be handled explicitly.
- Use `Response.json()` to create a JSON response.

## [0.0.4] - 2023-11-13
### Fixed
- Try to fix npm release

## [0.0.3] - 2023-11-13
### Fixed
- Try to fix npm release

## [0.0.2] - 2023-11-13
### Fixed
- Try to fix npm release

## [0.0.1] - 2023-11-13
### Added
- First release

## [0.0.0] - 2023-11-13
### Added
- First implementation

[Unreleased]: https://github.com/rcmachado/changelog/compare/v0.3.1...HEAD
[0.3.1]: https://github.com/rcmachado/changelog/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/rcmachado/changelog/compare/v0.2.1...v0.3.0
[0.2.1]: https://github.com/rcmachado/changelog/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/rcmachado/changelog/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/rcmachado/changelog/compare/v0.0.4...v0.1.0
[0.0.4]: https://github.com/rcmachado/changelog/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/rcmachado/changelog/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/rcmachado/changelog/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/rcmachado/changelog/compare/v0.0.0...v0.0.1
[0.0.0]: https://github.com/rcmachado/changelog/compare/dc9169d6918a38300db35e2e01d372e1c4f4d7d9...v0.0.0


src/index.ts
import type { ResponseConfig, RouteConfig } from '@asteasolutions/zod-to-openapi';
import type { ZodType } from 'zod';
import { type SafeParseReturnType, z } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';

export type FetchRouteContext<Params extends StringParams> = {
  params: Params;
  request: Request;
};

export type FetchRoute<Params extends StringParams = StringParams> = (
  context: FetchRouteContext<Params>,
) => Promise<Response>;

export type StringParams = Record<string, string | undefined>;

export type TypedParams = Record<string, string | number | undefined>;

type ObjectType = 'params' | 'query' | 'requestBody' | 'responseBody';

export class HttpError extends Error {
  public readonly response: Response;

  constructor(message: string, status: number) {
    super(message);
    this.response = errorResponse(message, status);
  }
}

export const ErrorSchema = z.object({
  message: z.string(),
});

export type UnwrappedError = {
  message: string;
  response: Response;
};

export function toHttpError(error: unknown): HttpError {
  if (error instanceof HttpError) {
    return error;
  } else if (error instanceof Error) {
    const httpError = new HttpError(error.message, 500);
    httpError.cause = error;
    return httpError;
  } else {
    const message = 'Unknown error';
    return new HttpError(message, 500);
  }
}

export const Response404: ResponseConfig = {
  description: 'Not Found',
  content: {
    'application/json': {
      schema: ErrorSchema,
    },
  },
};

export const Response415: ResponseConfig = {
  description: 'Unsupported Media Type',
  content: {
    'application/json': {
      schema: ErrorSchema,
    },
  },
};

export const Response422: ResponseConfig = {
  description: 'Unprocessable Entity',
  content: {
    'application/json': {
      schema: ErrorSchema,
    },
  },
};

export const Response500: ResponseConfig = {
  description: 'Internal Server Error',
  content: {
    'application/json': {
      schema: ErrorSchema,
    },
  },
};

export class Validator {
  // public readonly Response: typeof Response;

  constructor(private routeConfig: RouteConfig) {
    // this.Response = makeResponse(routeConfig);
  }

  params<T extends TypedParams>(params: StringParams): T {
    const schema = this.routeConfig.request?.params;
    return this.validateObject<T>(schema as unknown as ZodType<T>, params, 'params', 404);
  }

  query<T extends TypedParams>(request: Request): T {
    const schema = this.routeConfig.request?.query;
    const url = new URL(request.url, 'http://dummy.com');
    const query = Object.fromEntries(url.searchParams.entries());
    return this.validateObject<T>(schema as unknown as ZodType<T>, query, 'query', 404);
  }

  async body<T>(request: Request): Promise<T> {
    const contentType = request.headers.get('content-type');
    if (!contentType) {
      return null as T;
    }
    const schema = this.routeConfig.request?.body?.content[contentType]?.schema;
    if (!schema) {
      throw new HttpError(`No schema for Content-Type: ${contentType}`, 415);
    }
    if (contentType === 'application/json') {
      const body = await request.json();
      return this.validateObject<T>(schema as ZodType<T>, body, `requestBody`, 422);
    } else if (contentType === 'application/x-www-form-urlencoded') {
      const body = await request.formData();
      const data = Object.fromEntries(body.entries());
      return this.validateObject<T>(schema as ZodType<T>, data, `requestBody`, 422);
    }
    throw new HttpError(`Unsupported Content-Type: ${contentType}`, 415);
  }

  async validate(response: Response): Promise<Response> {
    const copy = response.clone();

    const status = copy.status;

    const responseConfig = this.routeConfig.responses[status];
    if (!responseConfig) {
      const statuses = Object.keys(this.routeConfig.responses).join(', ');
      return errorResponse(
        `No response config for status ${status}. Allowed statuses: ${statuses}`,
        500,
      );
    }

    const contentType = copy.headers.get('content-type');
    if (contentType === 'application/json') {
      if (!responseConfig.content) {
        return errorResponse(`No response config content for status ${status}`, 500);
      }
      const schema = responseConfig.content[contentType].schema;
      try {
        const data = await copy.json();
        this.validateObject(schema as ZodType<unknown>, data, 'responseBody', 500);
      } catch (error) {
        const { response } = toHttpError(error);
        return response;
      }
    }
    return response;
  }

  private validateObject<T>(
    schema: ZodType<T>,
    value: unknown,
    type: ObjectType,
    errorStatus: number,
  ): T {
    if (schema === undefined) {
      throw new HttpError(`No ${type} schema`, 500);
    }
    const result = schema.safeParse(value) as SafeParseReturnType<unknown, T>;
    if (result.success) {
      return result.data;
    }
    const valueJson = JSON.stringify(value, null, 2);
    const schemaJson = JSON.stringify(zodToJsonSchema(schema), null, 2);
    const errorMessage = `Error validating ${type}: ${valueJson}\n\nSchema: ${schemaJson}`;
    throw new HttpError(errorMessage, errorStatus);
  }
}

/**
 * Creates a Response with a { message } response body.
 * @param message the error message
 * @param status the HTTP status code
 * @returns a Response
 */
export function errorResponse(message: string, status: number): Response {
  return Response.json({ message }, { status });
}
